### PR TITLE
SONARPY-579 Rule S2638: Method overrides should not change contracts

### PR DESCRIPTION
**Pull request by andrea-guarino-sonarsource**
**DO NOT SQUASH**

### CODE REVIEW

File: python-checks/src/main/java/org/sonar/python/checks/ChangeMethodContractCheck.java [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+      }
+    }
+    return Optional.empty();
+  }
+
+
+  private static void reportOnMissingParameters(SubscriptionContext ctx, FunctionSymbol overriddenMethod, FunctionDef functionDef) {
+    int indexFirstMissingParam = TreeUtils.nonTupleParameters(functionDef).size();
+    StringBuilder missingParametersBuilder = new StringBuilder();
+    for (int i = indexFirstMissingParam; i < overriddenMethod.parameters().size(); i++) {
```
[the remark this answers is not in the document]
Comment (andrea-guarino-sonarsource) [reply python-checks/src/main/java/org/sonar/python/checks/ChangeMethodContractCheck.java]: Done in dd24cdb314b20f86ff1b404066898b087ceae2d3

File: python-checks/src/test/resources/checks/changeMethodContract.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+  def with_default(self, param1=42): ...
+  def changed_param_order(self, param1, param2): ...
+  def with_keyword_only(self, param1, *, param2): ...
+  def with_kwargs(self, param1, **kwargs): ...
+  @foo
+  def with_decorator(self, param1): ...
+  def __private_method(self): ...
+  attr = 42
+
+class ChildClass(ParentClass):
```
Comment (gyula-sallai-sonarsource) [inline python-checks/src/test/resources/checks/changeMethodContract.py]: I would also add a test for a case when a class inherits transitively from another.

File: python-frontend/src/main/java/org/sonar/python/semantic/SymbolTableBuilder.java [hunk 3]
```diff
@@ -155,7 +155,10 @@ private void addSymbolsToTree(FileInputImpl fileInput) {
           .filter(symbol -> symbol.kind() == Symbol.Kind.CLASS)
```
[the remark this answers is not in the document]
Comment (andrea-guarino-sonarsource) [reply python-frontend/src/main/java/org/sonar/python/semantic/SymbolTableBuilder.java]: Good point! done in f8d59ad6528d49ed17063123446871b22050921e

File: python-checks/src/test/resources/checks/changeMethodContract.py (hunk 2 above)
Comment (andrea-guarino-sonarsource) [reply python-checks/src/test/resources/checks/changeMethodContract.py]: Test added in eb3d4366f4418598816482ff85941ff7650a314f

**Pull request merged**